staging/src/k8s.io/component-base/config/testing/roundtrip.go

```diff
@@ -0,0 +1,119 @@
+package testing
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"testing"
+
+	"github.com/pkg/errors"
+	"github.com/pmezard/go-difflib/difflib"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/util/sets"
+	configserializer "k8s.io/component/pkg/serializer"
+)
+
+type TestCase struct {
+	name, in, out string
+	inGVK         schema.GroupVersionKind
+	outGV         schema.GroupVersion
+}
+
+func GetRoundtripTestCases(scheme *runtime.Scheme, disallowMarshalGroupVersions sets.String) []TestCase {
+	cases := []TestCase{}
+	versionsForKind := map[schema.GroupKind][]string{}
+	for gvk := range scheme.AllKnownTypes() {
+		versionsForKind[gvk.GroupKind()] = append(versionsForKind[gvk.GroupKind()], gvk.Version)
+	}
+
+	for gk, versions := range versionsForKind {
+		for _, vin := range versions {
```

Comment:
can we embeded it into line29, discard the struct `versionsForKind `?

Reply:
`versionsForKind` is necessary, because we need a complete list of versions contained within a particular GroupKind.

This is looping through versions twice in the same nested loop to generate all pairs of GroupKind related versions.